### DISCUSSION TITLE
Advanced integration specific configuration

### DISCUSSION THREAD

**Discussion by teharris1** [home-assistant/architecture#633] [General]
## Context

Many technologies require specific features or capabilities to be available to the user in order to fully utilize the technology. Z-Wave and ZigBee (ZHA) have solved this by having a panel under the configuration panel which allows for commands to be executed against the network or against specific devices. However, with the introduction of the device registry, the approach appears to moving toward integration specific controls on the device details page. This has been implemented for MQTT. 

An example is the Insteon component which requires a number of device-specific features to fully enable the technology such as:
1. Display device configuration such as dimming ramp rates, motion sensor time-outs, etc
1. Change the device configuration 
2. Load / reload a device's link database
3. Editing a device's all link database to clean up stale entries

It also has a set of network-level features that would need to be available: Examples are:
1. Creating a scene that triggers multiple devices to respond (i.e. TV watching scene).
2. Identify and fix broken links across all devices in the network
3. Remove dead devices from the network.

## Proposal

Create a standard way of deploying integration specific administration and configuration at a device and network level. This would include:
1. Creating a simplified way to add integration-specific commands to the device details page
2. Create an integration details card that will enable network (or integration wide) commands or configuration.

## Consequences

What becomes easier:
- Developers having a clean path to integration specific configuration that is more complex than config flows would make many integrations better for the user.
- Users having a clear path to get to integration specific configuration would make their life easier. 

More difficult to do and any risks introduced by the change that will need to be mitigated.
- Having two configuration paths, one for devices and one for networks may be confusing to users
- Z-Wave and ZHA would need to be refactored. Given they are key elements of HA today and very widely used, this may present risk.

**Comment by balloob** [comment]
Yes, a way of offering device options should definitely be added. Network level can be done using an options flow for the the config entry

**Comment by teharris1** [comment]
Options flow seems like a fairly limited UI today. If I look at the Z-Wave panel I would have to look if all of the network commands would be able to work as options flow. Would you consider that UI being expanded?  For some integrations, there is a Hub. Insteon, Z-Wave, and ZigBee all have a modem/hub in the network. Network-specific commands could go there.

**Comment by shbatm** [comment]
I've wondered about this for the isy994 side, which has the same points as above with Insteon.

In a forthcoming PR The ISY will be migrated to a config flow and the ISY itself will have a device added to the registry per [here (bottom of page](https://developers.home-assistant.io/docs/device_registry_index/#defining-devices)).

So if we add the device options, could we also just add the hub-level commands/options to the device that represents the hub?

**Comment by teharris1** [comment]
Here is a mock up of a potential integration specific configuration layout if we were to use the `Devices` info page for configuration options:
![image](https://user-images.githubusercontent.com/9055260/83933269-0c718580-a775-11ea-9947-0687c9c71cba.png)

The key elements here are:
1) The standard device info items of `Device info`, `Entities`, `Automations`, `Scenes` and `Scripts` would never change layout. 
2) Integration specific cards would be below the standard cards. This way we never disrupt the standard cards.
3) We should find an easy way for integration developers to add integration specific cards. Having each integration add integration specific code to `src/panels/config/devices/ha-config-device-page.ts` would be a nightmare.

**Comment by balloob** [comment]
This is what we're currently doing for MQTT already, and are in the process of adding Zigbee to this as well.

See also the ongoing config redesign mockup https://www.figma.com/file/4LZqQZRynDZGTgA2rN35Te/Home-Assistant-config-mockup?node-id=25%3A682

**Comment by teharris1** [comment]
Is it preferable to minimize the cards and changes on the device screen with links to detailed elements or is it better to have integration specific items on the device screen? I notice with the MQTT and the ZigBee design, the only change to the device page is a set of links to integration specific details. What I was thinking for Insteon was to extend the device page with integration specific cards. I can see benefits to both.  In favor of the links only approach this would minimize the code changes to `ha-config-device-page.ts` with an easier model to maintain. In favor of placing cards directly on the device page is the user-friendly access to information at the expense of a very large `ha-config-device-page.ts` file.  Alternatively, we can allow the integration owner make that call for now.

**Comment by balloob** [comment]
I don't think that we have fully decided yet on what path to take.

**Comment by teharris1** [comment]
Starting to develop an Insteon device configuration UI in earnest and looking for direction. I have been following the ZHA updates (albeit with out any ZigBee devices so...). :

The ZHA integration uses three main approaches:
  a. A configuration panel accessible via the `Integrations` page replacing the `Options` button with a `Configure` button. This appears to be for network wide items.
  b. Device specific info and actions accessible via the `Device` page using dialog windows
  c. Device specific info and actions accessible via the `Device` page using the configuration panel in `a` above. This appears to be for more complex device settings that cannot work using a dialog window.

Is this an agreed to approach? I am hoping to copy the model if at all possible.

**Comment by ziv1234** [comment]
I am trying to do the same for the Dynalite integration. I tried looking at the ZHA, but it is huge. Is there anywhere that this is explained? couldn't find the info and the frontend dev website, so was hoping that someone may have some pointers

**Comment by ziv1234** [comment]
UPDATE - the MQTT is a more simple example, so in the process of writing a config panel to Dynalite. I agree that a standardized way is better, but the question is how many integrations need complex options. As long as it is a flat dictionary for the config, options flow seems very simple and doesn't need any UI.

Is the proposal to have it all done via the backend? That would have been super useful as I had to learn TypeScript and Polymer to be able to write in the frontend...

**Comment by teharris1** [comment]
Perhaps this issue can be broken down into two subtopics:
1. Network / component level configuration which, in theory, can be handled via Config Flow.
2. Device specific configuration which does not have a meaningful solution today.

It seems to me it would make sense to have a Config Flow concept for devices. The main advantage of this concept would be to enable device specific configuration without frontend development. The question is how extensible this would need to be for all of the different component specific devices. My gut tells me you have a few types of device configuration concepts:
a. Individual settings  (i.e. max brightness, text to display on a device screen, timeout value, enable/disable etc.)
b. Grouped settings (thermostat built in schedule, database of connected devices, lock codes etc.)
c. Actions (refresh device settings, factory reset device, etc.)

Of course there may be more but all of those concepts feel like you could create a config flow function that would appear on the device details page to handle device configuration settings. 

There is also a question of how useful Config Flow is for network / component level configuration. MQTT, Z-Wave (OZW, Zwave, Zwave-JS) and ZHA all have custom network level configuration panels built into the front end. I am not sure if this was because of legacy reasons or if the Config Flow features have gaps to allow the work to be done in the Config Flow options only. @balloob I think you are intimately familiar with all of the components with custom panels. Do you believe you could replace the network level configuration with Config Flow options?

**Comment by frenck** [comment]
This architecture issue is old, stale, or possibly obsolete. Things changed a lot over the years.

For that reason, I'm going to close this discussion for now to clean up a little.

../Frenck

**Reply by ziv1234** [reply]
Makes sense. I believe it is still relevant, but perhaps the new ability to build custom panels can be a simpler solution for the few integrations where configuration is more complex than the normal config flow allows
